File: lib/src/main/scala/net/kemitix/thorp/lib/Counters.scala
package net.kemitix.thorp.lib

import net.kemitix.thorp.domain.SimpleLens

final case class Counters(
    uploaded: Int,
    deleted: Int,
    copied: Int,
    errors: Int
)

object Counters {
  val empty: Counters = Counters(0, 0, 0, 0)
  val uploaded: SimpleLens[Counters, Int] =
    SimpleLens[Counters, Int](_.uploaded, b => a => b.copy(uploaded = a))
  val deleted: SimpleLens[Counters, Int] =
    SimpleLens[Counters, Int](_.deleted, b => a => b.copy(deleted = a))
  val copied: SimpleLens[Counters, Int] =
    SimpleLens[Counters, Int](_.copied, b => a => b.copy(copied = a))
  val errors: SimpleLens[Counters, Int] =
    SimpleLens[Counters, Int](_.errors, b => a => b.copy(errors = a))
}

File: domain/src/main/java/net/kemitix/thorp/domain/Counters.java
package net.kemitix.thorp.domain;

import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;
import lombok.With;

@With
@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
public class Counters {
    public final int uploaded;
    public final int deleted;
    public final int copied;
    public final int errors;
    public static Counters empty() {
        return new Counters(0, 0, 0, 0);
    }
    public Counters incrementUploaded() {
        return withUploaded(uploaded + 1);
    }
    public Counters incrementDeleted() {
        return withDeleted(deleted + 1);
    }
    public Counters incrementCopied() {
        return withCopied(copied + 1);
    }
    public Counters incrementErrors() {
        return withErrors(errors + 1);
    }
}
